### ISSUE TITLE
False positive from clippy::needless_lifetimes

### ISSUE THREAD

**Issue by LilyFoote** [rust-lang/rust-clippy#13923]
### Summary

When converting from an enum with one lifetime to an enum with two lifetimes, I get a `needless_lifetimes` warning for the second lifetime that becomes an error when removed.

### Lint Name

needless_lifetimes

### Reproducer

I tried [this code](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e7ef1b4207e061bfadde10cdc5437af5):

```rust
struct Py<'py> {
    data: &'py str,
}

enum Content<'t, 'py> {
    Py(Py<'py>),
    T1(&'t str),
    T2(&'t str),
}

enum ContentString<'t> {
    T1(&'t str),
    T2(&'t str),
}

impl<'t, 'py> ContentString<'t> {
    fn map_content(self, f: impl FnOnce(&'t str) -> &'t str) -> Content<'t, 'py> {
        match self {
            Self::T1(content) => Content::T1(f(content)),
            Self::T2(content) => Content::T2(f(content)),
        }
    }
}
```

I saw this happen:

```
warning: the following explicit lifetimes could be elided: 'py
  --> src/main.rs:16:10
   |
16 | impl<'t, 'py> ContentString<'t> {
   |          ^^^
17 |     fn map_content(self, f: impl FnOnce(&'t str) -> &'t str) -> Content<'t, 'py> {
   |                                                                             ^^^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_lifetimes
   = note: `#[warn(clippy::needless_lifetimes)]` on by default
help: elide the lifetimes
   |
16 ~ impl<'t> ContentString<'t> {
17 ~     fn map_content(self, f: impl FnOnce(&'t str) -> &'t str) -> Content<'t, '_> {
   |
```

I expected to see no warning.

If I [remove the `'py` lifetime from `map_content`](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=f5ee1b2848444aa6d90485e2dfe448b3):

```rust
impl<'t> ContentString<'t> {
    fn map_content(self, f: impl FnOnce(&'t str) -> &'t str) -> Content<'t, '_> {
        match self {
            Self::T1(content) => Content::T1(f(content)),
            Self::T2(content) => Content::T2(f(content)),
        }
    }
}
```
I get a compiler error:

```
error[E0106]: missing lifetime specifier
  --> src/main.rs:17:77
   |
17 |     fn map_content(self, f: impl FnOnce(&'t str) -> &'t str) -> Content<'t, '_> {
   |                                                                             ^^ expected named lifetime parameter
   |
   = help: this function's return type contains a borrowed value, but there is no value for it to be borrowed from
help: consider using the `'t` lifetime
   |
17 |     fn map_content(self, f: impl FnOnce(&'t str) -> &'t str) -> Content<'t, 't> {
   |                                                                             ~~

For more information about this error, try `rustc --explain E0106`.
```

### Version

```text
rustc 1.83.0 (90b35a623 2024-11-26)
binary: rustc
commit-hash: 90b35a6239c3d8bdabc530a6a0816f7ff89a0aaf
commit-date: 2024-11-26
host: x86_64-unknown-linux-gnu
release: 1.83.0
LLVM version: 19.1.1
```

### Additional Labels

@rustbot label +I-suggestion-causes-error
